pytest_twisted.py

```diff
@@ -83,19 +93,55 @@ def stop_twisted_greenlet():
         _instances.gr_twisted.switch()
 
 
+def is_coroutine(something):
+    if ASYNC_AWAIT:
+        return asyncio.iscoroutine(something)
+
+    return False
+
+
+def is_async_generator(something):
+    if ASYNC_GENERATORS:
+        return inspect.isasyncgen(something)
+
+    return False
+
+
+@defer.inlineCallbacks
 def _pytest_pyfunc_call(pyfuncitem):
     testfunction = pyfuncitem.obj
+    async_generators = []
     if pyfuncitem._isyieldedfunction():
-        return testfunction(*pyfuncitem._args)
+        defer.returnValue(testfunction(*pyfuncitem._args))
     else:
         funcargs = pyfuncitem.funcargs
         if hasattr(pyfuncitem, "_fixtureinfo"):
             testargs = {}
             for arg in pyfuncitem._fixtureinfo.argnames:
-                testargs[arg] = funcargs[arg]
+                something = funcargs[arg]
+                if is_coroutine(something):
+                    something = yield defer.ensureDeferred(something)
+                elif is_async_generator(something):
+                    async_generators.append(something)
+                    something = yield defer.ensureDeferred(
+                        something.__anext__(),
+                    )
+                testargs[arg] = something
         else:
             testargs = funcargs
-        return testfunction(**testargs)
+        result = yield testfunction(**testargs)
+
+        for async_generator in async_generators:
+            try:
+                yield defer.ensureDeferred(
+                    async_generator.__anext__(),
+                )
```

Comment:
I wonder if there are cases where iterating all currently-outstanding generators here "at the same time" would be important? (Alternatively, maybe it's better to invoke them serially as here...)

Reply:
Yeah,  serially async...  :[  I'll fix that.